### DISCUSSION TITLE
How does the "short circuit evaluation of conditions" work?

### DISCUSSION THREAD

**Discussion by Tpal12** [WarmUpTill/SceneSwitcher#1241] [Q&A] [answered]
Hey, just for my understanding: What are the criteria to shortcut the evaluation?

Is it based on: at first node == false -> break & return 
Or another mechanism?

Thank you in advance and have a nice day :)

**Comment by WarmUpTill** [comment, marked as answer]
Hi,
There is no custom short circuit logic implementation done on my side, as I am simply relying on C++ doing the short circuit evaluation for me here.

If the current "condition evaluation result" of a macro cannot change with a given logical operator and the given input parameters the evaluation of a condition is skipped.

Let's assume you have the following conditions in a macro ...
![](https://github.com/user-attachments/assets/1e4522e5-3be3-488e-887f-749029f9f3ec)
... and the state of the conditions is:

* File => true
* Scene => false
* Timer => true

The conditions will be evaluated from top to bottom resulting in this expression having to evaluate to `true` for the macro to be executed:

`((File || Scene) && Timer)`

As the `File` check returns `true`, the `Scene` check will be skipped, because `File || Scene` will result in `true || Scene`, which will always evaluate to `true` regardless of the value of `Scene`.

The `Timer` check cannot be skipped as `true || Timer` might be `true` or `false` depending on the value of `Timer`.

I hope that helps!

**Reply by Tpal12** [reply]
This is actually helpful. 

So the difference is: 
when I activate short-circuit evaluation
All “or statements” are resolved in their own parenthetical evaluation (evaluation and return on the first true)
otherwise it is a complete evaluation of all || no matter what?

In that case, I see myself simplifying some of my macros and saving resources
By intentionally  checking the least resource consuming states first and hoping it´s already true ^^

Thank you for your answer, I think I understood

**Reply by WarmUpTill** [reply]
I am glad to hear that it was helpful! :)

I just have to add the following note:
This is not limited to the "or" operator, but a general setting affecting all logical operations.

So for example if you have the following conditions ...
![](https://github.com/user-attachments/assets/b76599fd-a3a4-4541-921a-5980bc11cae4)
... and the state of the conditions is:

* Scene => false
* Timer => (doesn't matter)

It would result in the expression `Scene && Timer`.
As Scene is `false` it can be written as `false && Timer`.
No matter the value of `Timer` the expression can never evaluate to `true` and will always be `false`.
Thus the `Timer` check is skipped.

**Reply by Tpal12** [reply]
Ohh, in this case my approach would lead to some sneaky side effects and inconsistencies.
That's a shame, but not the end of the world. 
I'll happily experiment with the possibilities and limitations once it leaves beta status

But thanks to you, I already have a basic understanding of the implementation itself and therefore an entry point to Google (C++ Short-circuit evaluation) to get deeper insights if needed

Kind regards, have a great evening and thank you very much :)
